kafka/src/main/java/org/axonframework/extensions/kafka/eventhandling/cloudevent/ExtensionUtils.java

```diff
@@ -0,0 +1,159 @@
+
+    private static final Set<String> NON_METADATA_EXTENSIONS = Stream
+            .of(AGGREGATE_TYPE, AGGREGATE_ID, AGGREGATE_SEQ, MESSAGE_REVISION)
+            .collect(Collectors.toCollection(HashSet::new));
+
+    private ExtensionUtils() {
+        // Utility class
+    }
+
+    static boolean isNonMetadataExtension(String extensionName) {
+        return NON_METADATA_EXTENSIONS.contains(extensionName);
+    }
+
+    static void setExtension(CloudEventBuilder builder, String extensionName, Object value) {
+        if (isNonMetadataExtension(extensionName)) {
+            throw new InvalidMetaDataException(
+                    String.format("Metadata property '%s' is already reserved to be used for Axon",
+                                  extensionName)
+            );
+        }
+        if (!isValidExtensionName(extensionName)) {
+            throw new InvalidMetaDataException(
+                    String.format("Metadata property '%s' is not a valid extension name",
+                                  extensionName)
+            );
+        }
+        if (value instanceof String) {
+            builder.withExtension(extensionName, (String) value);
+        } else if (value instanceof Number) {
+            builder.withExtension(extensionName, (Number) value);
+        } else if (value instanceof Boolean) {
+            builder.withExtension(extensionName, (Boolean) value);
+        } else if (value instanceof URI) {
+            builder.withExtension(extensionName, (URI) value);
+        } else if (value instanceof OffsetDateTime) {
+            builder.withExtension(extensionName, (OffsetDateTime) value);
+        } else if (value instanceof byte[]) {
+            builder.withExtension(extensionName, (byte[]) value);
+        } else {
+            throw new InvalidMetaDataException(
+                    String.format("Metadata property '%s' is of class '%s' and thus can't be added",
+                                  extensionName,
+                                  value.getClass())
+            );
+        }
+    }
+
+    static String asNullableString(Object object) {
+        if (object instanceof String) {
+            return (String) object;
+        } else {
+            return null;
+        }
+    }
+
+    static Long asLong(Object object) {
+        if (object instanceof Long) {
+            return (Long) object;
+        } else {
+            return 0L;
+        }
+    }
+
+    static OffsetDateTime asOffsetDateTime(Object object, long fallbackTimestamp) {
+        if (object instanceof OffsetDateTime) {
+            return (OffsetDateTime) object;
+        } else {
+            return Instant.ofEpochMilli(fallbackTimestamp).atOffset(ZoneOffset.UTC);
+        }
+    }
+
+    static byte[] asBytes(CloudEventData data) {
+        if (isNull(data)) {
+            return new byte[0];
+        } else {
+            return data.toBytes();
+        }
+    }
+
+    static boolean isValidExtensionName(String name) {
```

Comment:
I think we can solve this with a Pattern/regex:

```java
Pattern pattern = Pattern.compile("^([0-9]|[a-z])+$");
return pattern.matcher(name).matches();
```

Reply:
I rather use the exact method that's used in Cloud Events. Although a regex might be more efficient. There also seems to be a max length. But the current Java code isn't validating it.

Reply:
liked the idea, so opened a PR to change in CouldEvents. we could also use that here. The max length is a should and not a must. @smcvb do you have a preference for using the same, or switching to regex? https://github.com/cloudevents/sdk-java/pull/477